File: go/astraeadb/internal/backoff/backoff_test.go
package backoff

import (
	"testing"
	"time"
)

func TestDefaultBackoff(t *testing.T) {
	b := Default()
	if b.min != 100*time.Millisecond {
		t.Errorf("min = %v, want 100ms", b.min)
	}
	if b.max != 10*time.Second {
		t.Errorf("max = %v, want 10s", b.max)
	}
}

func TestBackoffIncreases(t *testing.T) {
	b := New(100*time.Millisecond, 10*time.Second, 2.0)

	// First call should be around 100ms (with jitter).
	d1 := b.Next()
	if d1 < 75*time.Millisecond || d1 > 125*time.Millisecond {
		t.Errorf("first delay = %v, expected ~100ms", d1)
	}

	// Second call should be around 200ms (with jitter).
	d2 := b.Next()
	if d2 < 150*time.Millisecond || d2 > 250*time.Millisecond {
		t.Errorf("second delay = %v, expected ~200ms", d2)
	}
}

func TestBackoffCapsAtMax(t *testing.T) {
	b := New(1*time.Second, 2*time.Second, 10.0)

	// First call: ~1s
	b.Next()
	// Second call should cap at max (~2s with jitter).
	d := b.Next()
	if d > 3*time.Second {
		t.Errorf("delay = %v, should not exceed max by much", d)
	}
}

func TestBackoffReset(t *testing.T) {
	b := New(100*time.Millisecond, 10*time.Second, 2.0)

	b.Next()
	b.Next()
	b.Reset()

	d := b.Next()
	if d < 75*time.Millisecond || d > 125*time.Millisecond {
		t.Errorf("after reset, delay = %v, expected ~100ms", d)
	}
}

File: java/astraeadb/astraeadb-json/src/test/java/com/astraeadb/json/BackoffTest.java
package com.astraeadb.json;

import org.junit.jupiter.api.Test;

import java.time.Duration;

import static org.assertj.core.api.Assertions.assertThat;

class BackoffTest {

    @Test
    void defaultValues() {
        ExponentialBackoff backoff = new ExponentialBackoff();
        Duration d = backoff.nextDelay();
        // 100ms with +/-25% jitter → 75ms to 125ms
        assertThat(d.toMillis()).isBetween(75L, 125L);
    }

    @Test
    void exponentialIncrease() {
        ExponentialBackoff backoff = new ExponentialBackoff();
        long first = backoff.nextDelay().toMillis();
        long second = backoff.nextDelay().toMillis();
        long third = backoff.nextDelay().toMillis();
        // The base doubles each time: 100 → 200 → 400
        // With jitter, second should generally be larger than first
        // We check the third is larger than the initial base to confirm growth
        assertThat(third).isGreaterThan(first);
    }

    @Test
    void maxCap() {
        ExponentialBackoff backoff = new ExponentialBackoff();
        // Call many times to reach the cap
        Duration last = Duration.ZERO;
        for (int i = 0; i < 50; i++) {
            last = backoff.nextDelay();
        }
        // 30_000ms cap with 25% jitter → max possible is 37500ms
        assertThat(last.toMillis()).isLessThanOrEqualTo(37_500L);
        // Base should be capped at 30000, so with -25% jitter, min is 22500
        assertThat(last.toMillis()).isGreaterThanOrEqualTo(22_500L);
    }

    @Test
    void reset() {
        ExponentialBackoff backoff = new ExponentialBackoff();
        // Advance several times
        for (int i = 0; i < 10; i++) {
            backoff.nextDelay();
        }
        backoff.reset();
        Duration d = backoff.nextDelay();
        // After reset, should be back to ~100ms (75-125ms with jitter)
        assertThat(d.toMillis()).isBetween(75L, 125L);
    }
}
